syne_tune/callbacks/tensorboard_callback.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+import numbers
+import os
+from time import perf_counter
+from typing import Optional, List
+
+from syne_tune.backend.trial_status import Trial
+from syne_tune.constants import ST_TUNER_TIME
+from syne_tune.tuner_callback import TunerCallback, logger
+
+
+class TensorboardCallback(TunerCallback):
+    """
+    Logs relevant metrics reported from trial evaluations, so they can be
+    visualized with Tensorboard.
+    """
+
+    def __init__(
+        self,
+        ignore_metrics: Optional[List[str]] = None,
+        target_metric: Optional[str] = None,
+        mode: Optional[str] = None,
+    ):
+        """
+        :param ignore_metrics: Defines which metrics should be ignored. If None,
+            all metrics are reported to Tensorboard.
+        :param target_metric: Defines the metric we aim to optimize. If this
+            argument is set, we report the cumulative optimum of this metric as
+            well as the optimal hyperparameters we have found so far.
+        :param mode: Determined whether we maximize ("max") or minimize ("min")
+            the target metric.
+        """
+        if mode is None:
+            mode = "min"
```

Comment:
I think it is best practice to define default values of base class methods in the code (so always use = None), otherwise they need to be copy and pasted to subclasses, and this introduces errors. Also, strings are mutable.

Reply:
I see your point about inheritance, as long as we keep the default in the docstrings then I am ok with that. 

I think the best practice are arguable, see for example an implementation of Adam in tensorflow https://github.com/keras-team/keras/blob/v2.10.0/keras/optimizers/optimizer_v2/adam.py#L26-L270 or sgd in Pytorch https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html#SGD which are(IMHO) a relatively well written library 

As a note, strings in python are immutable, see second sentence of https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str

Reply:
I just checked the sphinx docs, and they recommend to mention type and default in the docstring, something we are not right now doing. In the second PR, I will do that for all classes which are user-facing. This way, we get all of that info also in the API docs.